src/components/modules/blockEvents.ts

```diff
@@ -277,104 +282,166 @@ export default class BlockEvents extends Module {
    */
-  private mergeBlocks(): void {
-    const { BlockManager, Caret, Toolbar } = this.Editor;
-    const targetBlock = BlockManager.previousBlock;
-    const blockToMerge = BlockManager.currentBlock;
+  private delete(event: KeyboardEvent): void {
+    const { BlockManager, Caret } = this.Editor;
+    const { currentBlock, nextBlock } = BlockManager;
 
     /**
-     * Blocks that can be merged:
-     * 1) with the same Name
-     * 2) Tool has 'merge' method
-     *
-     * other case will handle as usual ARROW LEFT behaviour
+     * If some fragment is selected, leave native behaviour
      */
-    if (blockToMerge.name !== targetBlock.name || !targetBlock.mergeable) {
-      /** If target Block doesn't contain inputs or empty, remove it */
-      if (targetBlock.inputs.length === 0 || targetBlock.isEmpty) {
-        BlockManager.removeBlock(BlockManager.currentBlockIndex - 1);
+    if (!SelectionUtils.isCollapsed) {
+      return;
+    }
 
-        Caret.setToBlock(BlockManager.currentBlock);
-        Toolbar.close();
+    /**
+     * If caret is not at the end, leave native behaviour
+     */
+    if (!Caret.isAtEnd) {
+      return;
+    }
 
-        return;
-      }
+    /**
+     * All the cases below have custom behaviour, so we don't need a native one
+     */
+    event.preventDefault();
+    this.Editor.Toolbar.close();
 
-      if (Caret.navigatePrevious()) {
-        Toolbar.close();
-      }
+    const isLastInputFocused = currentBlock.currentInput === currentBlock.lastInput;
+
+    /**
+     * For example, caret at the end of the Quote first input (quote text) — just navigate next input (caption)
+     */
+    if (!isLastInputFocused) {
+      Caret.navigateNext();
 
       return;
     }
 
+    /**
+     * If next Block is empty, it should be removed just like a character
+     */
+    if (nextBlock.isEmpty) {
+      BlockManager.removeBlock(nextBlock);
+
+      return;
+    }
+
+    /**
+     * If current Block is empty, just just remove it and set cursor to the next Block (like we're removing line break char)
+     */
+    if (currentBlock.isEmpty) {
+      BlockManager.removeBlock(currentBlock);
+
+      /**
+       * Next Block is now current
+       */
+      Caret.setToBlock(BlockManager.currentBlock, Caret.positions.START);
+
+      return;
+    }
+
+    const bothBlocksMergeable = currentBlock.mergeable && nextBlock.mergeable;
+
+    /**
+     * If Blocks could be merged, do it
+     * Otherwise, just navigate previous block
+     */
+    if (bothBlocksMergeable) {
+      this.mergeBlocks(nextBlock, currentBlock);
+    } else {
+      Caret.setToBlock(
+        BlockManager.nextBlock,
+        Caret.positions.START
+      );
+    }
+  }
+
+  /**
+   * Merge current and previous Blocks if they have the same type
+   *
+   * @param blockToMerge - what Block we want to merge
+   * @param targetBlock - to which Block we want to merge
+   */
+  private mergeBlocks(blockToMerge: Block, targetBlock: Block): void {
```

Comment:
I think it's more clear to have a first argument as a target, and a second argument as data to merge it to the target block (as in the BlockManager.mergeBlocks)

```suggestion
  private mergeBlocks(targetBlock: Block, blockToMerge: Block): void {
```